### PR TITLE
[Core] Push message to driver when a Raylet dies

### LINKED ISSUE #24234: [Core][Nightly-test] scheduling_test_many_0s_tasks_many_nodes failed
### What happened + What you expected to happen

https://console.anyscale.com/o/anyscale-internal/projects/prj_2xR6uT6t7jJuu1aCwWMsle/clusters/ses_ZSSKSXJ5pjHAcws5g3Z2eGgM?command-history-section=command_history

```
[2m[33m(raylet, ip=172.31.70.203)[0m [2022-04-24 20:11:54,239 E 133 133] (raylet) worker_pool.cc:502: Some workers of the worker process(2377) have not registered within the timeout. The process is still alive, probably it's hanging during start.
[2m[33m(raylet, ip=172.31.67.141)[0m [2022-04-24 20:11:54,373 E 133 133] (raylet) worker_pool.cc:502: Some workers of the worker process(2521) have not registered within the timeout. The process is still alive, probably it's hanging during start.
...
[2m[33m(raylet, ip=172.31.66.189)[0m [2022-04-24 20:12:46,747 E 133 133] (raylet) worker_pool.cc:502: Some workers of the worker process(4031) have not registered within the timeout. The process is still alive, probably it's hanging during start.
[2m[33m(raylet, ip=172.31.66.189)[0m [2022-04-24 20:12:46,756 E 133 133] (raylet) worker_pool.cc:502: Some workers of the worker process(4032) have not registered within the timeout. The process is still alive, probably it's hanging during start.
[2m[33m(raylet, ip=172.31.66.189)[0m [2022-04-24 20:12:46,761 E 133 133] (raylet) worker_pool.cc:502: Some workers of the worker process(4033) have not registered within the timeout. The process is still alive, probably it's hanging during start.
[2m[33m(raylet, ip=172.31.66.189)[0m [2022-04-24 20:12:46,768 E 133 133] (raylet) worker_pool.cc:502: Some workers of the worker process(4034) have not registered within the timeout. The process is still alive, probably it's hanging during start.
[2m[33m(raylet, ip=172.31.66.189)[0m [2022-04-24 20:12:46,771 E 133 133] (raylet) worker_pool.cc:502: Some workers of the worker process(4035) have not registered within the timeout. The process is still alive, probably it's hanging during start.
[2m[33m(raylet, ip=172.31.66.189)[0m [2022-04-24 20:12:46,775 E 133 133] (raylet) worker_pool.cc:502: Some workers of the worker process(4036) have not registered within the timeout. The process is still alive, probably it's hanging during start.
[2m[33m(raylet, ip=172.31.69.48)[0m [2022-04-24 20:12:47,093 E 133 133] (raylet) worker_pool.cc:502: Some workers of the worker process(3486) have not registered within the timeout. The process is still alive, probably it's hanging during start.
[2m[33m(raylet, ip=172.31.69.48)[0m [2022-04-24 20:12:47,098 E 133 133] (raylet) worker_pool.cc:502: Some workers of the worker process(3487) have not registered within the timeout. The process is still alive, probably it's hanging during start.
Traceback (most recent call last):
  File "distributed/test_scheduling.py", line 105, in <module>
    args.total_num_actors, args.num_actors_per_nodes, job
  File "distributed/test_scheduling.py", line 53, in start_actor
    lambda: ray.get([actor.do_job.remote() for actor in actors])
  File "distributed/test_scheduling.py", line 38, in measure
    ret = f()
  File "distributed/test_scheduling.py", line 53, in <lambda>
    lambda: ray.get([actor.do_job.remote() for actor in actors])
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/worker.py", line 1827, in get
    raise value
ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.
```

### Versions / Dependencies

lastet ray

### Reproduction script

run nightly test

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by mwtian**
## Why are these changes needed?
Currently when Raylets die, it is hard to figure out:
- if a Raylet died at all in a cluster. Usually we have to check on nodes where a number of workers died and see if the Raylet has died as well.
- reason of Raylet's death.

With this PR, if a Raylet dies from a reason other than SIGTERM, the dashboard agent will report the failure along with last 20 lines of the Raylet log.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/agent.py [hunk 1]
```diff
@@ -152,7 +157,42 @@ async def _check_parent():
                 while True:
                     parent = curr_proc.parent()
                     if parent is None or parent.pid == 1 or self.ppid != parent.pid:
-                        logger.error("Raylet is dead, exiting.")
+                        log_path = "/tmp/ray/session_latest/logs/raylet.out"
+                        error = False
+                        msg = f"Raylet is terminated: ip={self.ip}, id={self.node_id}. "
+                        try:
+                            with open(log_path, "r") as f:
+                                raylet_logs = f.readlines()
```
Comment (rkooo567) [inline dashboard/agent.py:165]: Are we reading the whole file? Maybe we can read last X lines here as well?

File: dashboard/agent.py [hunk 2]
```diff
@@ -152,7 +157,42 @@ async def _check_parent():
                 while True:
                     parent = curr_proc.parent()
                     if parent is None or parent.pid == 1 or self.ppid != parent.pid:
-                        logger.error("Raylet is dead, exiting.")
+                        log_path = "/tmp/ray/session_latest/logs/raylet.out"
```
Comment (rkooo567) [inline dashboard/agent.py:160]: I think we should not read session_latest. Maybe passing the session dir properly here?

File: dashboard/agent.py (hunk 1 above)
Comment (mwtian) [reply dashboard/agent.py:179]: Sure, reading last 1MB if the log file is larger than that.

File: dashboard/agent.py (hunk 2 above)
Comment (mwtian) [reply dashboard/agent.py]: Sounds good.

File: dashboard/agent.py (hunk 1 above)
Comment (rkooo567) [reply dashboard/agent.py:180]: Can you add a test for this new update?

**Review by rkooo567 — COMMENTED** [review]
LGTM if tests for the new part is added!

File: dashboard/agent.py (hunk 1 above)
Comment (mwtian) [reply dashboard/agent.py:180]: sure

**Comment by mwtian** [comment]
Failures look unrelated.

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
Very excited for this feature!

**Pull request merged**